File: apps/android/PymesHub/app/src/main/java/com/pymeshub/android/core/api/ApiClient.kt
package com.pymeshub.android.core.api

import com.pymeshub.android.app.AuthProvider
import com.pymeshub.android.core.model.AuthResponse
import com.pymeshub.android.core.model.AuthUser
import com.pymeshub.android.core.model.Contact
import com.pymeshub.android.core.model.Conversation
import com.pymeshub.android.core.model.Invoice
import com.pymeshub.android.core.model.Message
import com.pymeshub.android.core.model.NotificationItem
import com.pymeshub.android.core.model.RefreshResponse
import com.pymeshub.android.core.model.TaskItem
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

class ApiClient(private val baseUrl: String) {
    var authProvider: AuthProvider? = null

    suspend fun login(email: String, password: String, workspaceSlug: String): AuthResponse {
        val body = JSONObject()
            .put("email", email)
            .put("password", password)
        val json = request(
            method = "POST",
            path = "/api/auth/login",
            body = body,
            headers = mapOf("x-workspace-slug" to workspaceSlug),
            requiresAuth = false
        )
        return AuthResponse.fromJson(JSONObject(json))
    }

    suspend fun refresh(refreshToken: String): RefreshResponse {
        val json = request(
            method = "POST",
            path = "/api/auth/refresh",
            body = JSONObject().put("refresh_token", refreshToken),
            requiresAuth = false
        )
        return RefreshResponse.fromJson(JSONObject(json))
    }

    suspend fun getMe(): AuthUser = AuthUser.fromJson(JSONObject(request("GET", "/api/auth/me")))

    suspend fun logout() {
        request("POST", "/api/auth/logout")
    }

    suspend fun getConversations(): List<Conversation> {
        return parseArray(request("GET", "/api/conversations")).map(Conversation::fromJson)
    }

    suspend fun getMessages(conversationId: String): List<Message> {
        return parseArray(request("GET", "/api/conversations/$conversationId/messages")).map(Message::fromJson)
    }

    suspend fun sendMessage(conversationId: String, text: String): Message {
        val body = JSONObject().put("body_text", text)
        return Message.fromJson(JSONObject(request("POST", "/api/conversations/$conversationId/messages", body)))
    }

    suspend fun resolveConversation(id: String) {
        request("POST", "/api/conversations/$id/resolve")
    }

    suspend fun getContacts(): List<Contact> {
        return parseArray(request("GET", "/api/contacts")).map(Contact::fromJson)
    }

    suspend fun getTasks(): List<TaskItem> {
        return parseArray(request("GET", "/api/tasks")).map(TaskItem::fromJson)
    }

    suspend fun completeTask(id: String) {
        request("POST", "/api/tasks/$id/complete")
    }

    suspend fun getInvoices(): List<Invoice> {
        return parseArray(request("GET", "/api/invoices")).map(Invoice::fromJson)
    }

    suspend fun getNotifications(): List<NotificationItem> {
        return parseArray(request("GET", "/api/notifications")).map(NotificationItem::fromJson)
    }

    suspend fun getUnreadNotificationCount(): Int {
        val json = JSONObject(request("GET", "/api/notifications/unread-count"))
        return json.optInt("count", 0)
    }

    suspend fun markAllNotificationsRead(): Int {
        val json = JSONObject(request("POST", "/api/notifications/mark-read", JSONObject()))
        return json.optInt("count", 0)
    }

    private suspend fun request(
        method: String,
        path: String,
        body: JSONObject? = null,
        headers: Map<String, String> = emptyMap(),
        requiresAuth: Boolean = true,
        didRetry: Boolean = false
    ): String = withContext(Dispatchers.IO) {
        val connection = URL("${baseUrl.trimEnd('/')}/${path.trimStart('/')}").openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.setRequestProperty("Accept", "application/json")
        headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }

        if (requiresAuth) {
            authProvider?.accessToken?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
            authProvider?.workspaceSlug?.let { connection.setRequestProperty("x-workspace-slug", it) }
        }

        if (body != null) {
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
        }

        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        connection.disconnect()

        if (status == 401 && requiresAuth && !didRetry && authProvider?.refreshAccessToken() == true) {
            return@withContext request(method, path, body, headers, requiresAuth, didRetry = true)
        }

        if (status !in 200..299) {
            if (status == 401) authProvider?.clearSession()
            throw ApiException(status, response.ifBlank { "HTTP $status" })
        }

        response.ifBlank { "{}" }
    }

    private fun parseArray(raw: String): List<JSONObject> {
        val trimmed = raw.trim()
        val array = if (trimmed.startsWith("[")) {
            JSONArray(trimmed)
        } else {
            val root = JSONObject(trimmed.ifBlank { "{}" })
            when {
                root.optJSONArray("data") != null -> root.getJSONArray("data")
                root.optJSONArray("items") != null -> root.getJSONArray("items")
                root.optJSONArray("results") != null -> root.getJSONArray("results")
                else -> JSONArray()
            }
        }

        return List(array.length()) { index -> array.getJSONObject(index) }
    }
}

File: apps/ios/PymesHub/PymesHub/Core/API/APIClient.swift
import Foundation

@MainActor
final class APIClient {
    let baseURL: URL
    weak var authProvider: AuthProviding?

    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
    }

    func login(email: String, password: String, workspaceSlug: String) async throws -> AuthResponse {
        let headers = ["x-workspace-slug": workspaceSlug]
        return try await send(
            method: "POST",
            path: "/api/auth/login",
            body: LoginRequest(email: email, password: password),
            headers: headers,
            requiresAuth: false
        )
    }

    func refresh(refreshToken: String) async throws -> RefreshResponse {
        try await send(
            method: "POST",
            path: "/api/auth/refresh",
            body: RefreshRequest(refreshToken: refreshToken),
            requiresAuth: false
        )
    }

    func getMe() async throws -> AuthUser {
        try await send(method: "GET", path: "/api/auth/me")
    }

    func logout() async throws {
        let _: EmptyResponse = try await send(method: "POST", path: "/api/auth/logout")
    }

    func getConversations() async throws -> [Conversation] {
        try await send(method: "GET", path: "/api/conversations")
    }

    func getMessages(conversationID: String) async throws -> [Message] {
        try await send(method: "GET", path: "/api/conversations/\(conversationID)/messages")
    }

    func sendMessage(conversationID: String, text: String) async throws -> Message {
        try await send(
            method: "POST",
            path: "/api/conversations/\(conversationID)/messages",
            body: SendMessageRequest(bodyText: text)
        )
    }

    func resolveConversation(id: String) async throws {
        let _: EmptyResponse = try await send(method: "POST", path: "/api/conversations/\(id)/resolve")
    }

    func getContacts() async throws -> [Contact] {
        try await send(method: "GET", path: "/api/contacts")
    }

    func getTasks() async throws -> [TaskItem] {
        try await send(method: "GET", path: "/api/tasks")
    }

    func completeTask(id: String) async throws {
        let _: EmptyResponse = try await send(method: "POST", path: "/api/tasks/\(id)/complete")
    }

    func getInvoices() async throws -> [Invoice] {
        try await send(method: "GET", path: "/api/invoices")
    }

    private func send<Response: Decodable>(
        method: String,
        path: String,
        requiresAuth: Bool = true
    ) async throws -> Response {
        try await send(method: method, path: path, body: Optional<String>.none, requiresAuth: requiresAuth)
    }

    private func send<Request: Encodable, Response: Decodable>(
        method: String,
        path: String,
        body: Request?,
        headers: [String: String] = [:],
        requiresAuth: Bool = true,
        didRetry: Bool = false
    ) async throws -> Response {
        let requestURL = url(for: path)
        var request = URLRequest(url: requestURL)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        if let body {
            request.httpBody = try encoder.encode(body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        headers.forEach { key, value in
            request.setValue(value, forHTTPHeaderField: key)
        }

        if requiresAuth {
            if let token = authProvider?.accessToken {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            if let workspaceSlug = authProvider?.workspaceSlug {
                request.setValue(workspaceSlug, forHTTPHeaderField: "x-workspace-slug")
            }
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        if http.statusCode == 401, requiresAuth, !didRetry, await authProvider?.refreshAccessToken() == true {
            return try await send(
                method: method,
                path: path,
                body: body,
                headers: headers,
                requiresAuth: requiresAuth,
                didRetry: true
            )
        }

        guard (200..<300).contains(http.statusCode) else {
            let message = String(data: data, encoding: .utf8) ?? HTTPURLResponse.localizedString(forStatusCode: http.statusCode)
            if http.statusCode == 401 {
                authProvider?.clearSession()
            }
            throw APIError.http(status: http.statusCode, message: message)
        }

        if data.isEmpty, Response.self == EmptyResponse.self {
            return EmptyResponse() as! Response
        }

        do {
            return try decoder.decode(Response.self, from: data)
        } catch {
            throw APIError.decoding(error)
        }
    }

    private func url(for path: String) -> URL {
        let normalizedBase = baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let normalizedPath = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return URL(string: "\(normalizedBase)/\(normalizedPath)")!
    }
}

struct EmptyResponse: Codable {}
